### PR TITLE
Skip lint for assertion evaluated to false

### LINKED ISSUE #119908: rustc is unable to evaluate a singular conditional pathway predicated on the length of an array literal
**rustc is unable to evaluate a singular conditional pathway predicated on the length of an array literal.**

This behavior prevents an obviously correct code to compile
(conditional should be eliminated altogether, instead of throwing a compile error in this case):
```Rust
fn main() {
    let x: [i32; 0] = [];
    if !x.is_empty() {
        x[0];
    }
}
```
thanks @compiler-errors 👍🏼 

[Reproducible example playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=db3135f7508eb61f4b54adfdbcd415d7).

### Meta
`rustc --version --verbose`:
```
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: aarch64-apple-darwin
release: 1.75.0
LLVM version: 17.0.6
```
(`stable`, `beta` & `nightly` are all affected seemingly independent of the platform or edition, incl. `2024`)

### Error output

```
Compiling playground v0.0.1 (/playground)
error: this operation will panic at runtime
 --> src/main.rs:4:9
  |
4 |         x[0];
  |         ^^^^ index out of bounds: the length is 0 but the index is 0
  |
  = note: `#[deny(unconditional_panic)]` on by default

error: could not compile `playground` (bin "playground") due to previous error
```

### PR DESCRIPTION
**Pull request by barabadzhi**
fixes #119908

### CODE REVIEW

File: tests/ui/const_prop/issue-119908.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// run-pass
+fn main() {
+    let x: [i32; 0] = [];
+    if x.len() > 0 {
+        x[0];
```
Comment (matthiaskrgr) [inline tests/ui/const_prop/issue-119908.rs:5]: does this trigger an unreachable/dead code code warning?

**Comment by oli-obk** [comment]
Your change disables all lints on slice length check assertions. I don't think this issue can easily be fixed, although we can special case `[T]::len` to be something the const propagator understands

**Comment by oli-obk** [comment]
Note that such a change will be much simpler to do after https://github.com/rust-lang/rust/pull/119627 lands

File: tests/ui/const_prop/issue-119908.rs (hunk 1 above)
Comment (barabadzhi) [reply tests/ui/const_prop/issue-119908.rs:5]: Nope, lint is omitted for the whole block.

**Comment by barabadzhi** [comment]
Yep, I see. Thanks @oli-obk.

I've done some more testing & it's the unwanted side effect indeed.
It's my first `rustc` PR, so I'm nowhere near fluent with the internals.

I'll wait for the #119627 to get merged & will revisit this PR.

**Comment by oli-obk** [comment]
The other PR has landed. Thinking about it a bit more, we may not need a special value at all. You could implemenent an arm for https://github.com/rust-lang/rust/blob/5bd5d214effd494f4bafb29b3a7a2f6c2070ca5c/compiler/rustc_mir_transform/src/const_prop_lint.rs#L820 and use a diagnostic item on the `[T]::len` method to detect that it's a call to that. If it is, just check the `self` type and extract the length in case it is an array. There may be some casts involved, so you may also need to implement some sort of helper for unsizing casts (which could probably be a PR on its own, but not sure we can write a test that will actually be affected).

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Pull request closed without merging**